### PR TITLE
Merge changes from cr_takeQuiz to stage branch

### PR DESCRIPTION
**Pull request by cindyryoo7**
Major changes since last PR:

- Added functionality to show global percentile after submitting a quiz
- Added functionality to show user's ranking amongst friends after submitting a quiz

Let me know if you have any questions or concerns. Thank you!

### CODE REVIEW

File: client/components/takeQuiz/TakeQuiz.jsx [hunk 1]
```diff
@@ -101,6 +107,8 @@ const TakeQuiz = (userId) => {
       return currentScore;
     })()
     .then(response => {
+      getFriendRankings(response);
+      getGlobalRankings(response);
       submitAnswers(response);
```
Comment (john-cooke832) [inline client/components/takeQuiz/TakeQuiz.jsx:112]: Slightly knit-picky, but can you be a little more explicit with what the response here is?

File: client/components/takeQuiz/TakeQuiz.jsx [hunk 2]
```diff
@@ -177,12 +185,103 @@ const TakeQuiz = (userId) => {
   }
 
   const submitAnswers = (userScore) => {
-    submitQuizAnswers(14, 1, userScore) // later change to: submitQuizAnswers(quizId, userId, userScore);
+    submitQuizAnswers({
+      correct_answer_count: userScore.correct,
+      incorrect_answer_count: userScore.incorrect,
+      id_quiz: quizId,
+      id_users: 11 // later change to: userId
```
Comment (john-cooke832) [inline client/components/takeQuiz/TakeQuiz.jsx:192]: Does this still need to use hard coded data?

File: client/components/takeQuiz/TakeQuiz.jsx [hunk 3]
```diff
@@ -177,12 +185,103 @@ const TakeQuiz = (userId) => {
...
       console.error('Error: cannot submit quiz answers to database', err);
     })
   };
 
+  const getGlobalRankings = (userScore) => {
+    let rankings = [], allScores;
+    let percentage = (userScore.correct / userScore.total * 100).toFixed(0);
+    getQuizGlobalRankings(quizId)
+      .then(response => {
+        allScores = response.data.rows;
```
Comment (john-cooke832) [inline client/components/takeQuiz/TakeQuiz.jsx:204]: Again, Kim and Tom told me to be a little more specific with what the response should actually be. In this case, it seems to be scores?

File: client/components/takeQuiz/TakeQuiz.jsx [hunk 4]
```diff
@@ -177,12 +185,103 @@ const TakeQuiz = (userId) => {
...
+          return [];
+        } else {
+          return allScores;
+        }
+      })
+      .then(response => {
+        if (response.length) {
+          for (let i = 0; i < response.length; i++) {
+            const correct = Number(response[i].correct_answer_count);
+            const incorrect = Number(response[i].incorrect_answer_count);
```
Comment (john-cooke832) [inline client/components/takeQuiz/TakeQuiz.jsx:216]: Again, responses

File: client/components/takeQuiz/TakeQuiz.jsx [hunk 5]
```diff
@@ -177,12 +185,103 @@ const TakeQuiz = (userId) => {
...
+          }
+          setGlobalScores(rankings);
+          return rankings;
+        } else {
+          return [];
+        }
+      })
+      .then(response => {
+        if (response.length) {
+          let sorted = response.sort();
```
Comment (john-cooke832) [inline client/components/takeQuiz/TakeQuiz.jsx:228]: Again

File: client/components/takeQuiz/TakeQuiz.jsx [hunk 6]
```diff
@@ -177,12 +185,103 @@ const TakeQuiz = (userId) => {
...
+          }
+          rankings.push(percentage);
+          let sorted = rankings.sort((a, b) => (b - a));
+          return sorted;
+        }
+      })
+      .then(response => {
+        if (response) {
+          setFriendScores(response);
+          let index = response.indexOf(percentage) + 1;
```
Comment (john-cooke832) [inline client/components/takeQuiz/TakeQuiz.jsx:275]: Again, response

File: server/index.js [hunk 7]
```diff
@@ -188,6 +188,39 @@ app.get(`/quiz/history/taken/:userId`, async (req, res) => {
...
 })
 
+app.get(`/quiz/rankings/global/:quizId`, async (req, res) => {
+  try {
+    const quizId = req.params.quizId;
+    const retrieveQuizGlobalRankings = await pool.query(
+      `SELECT * FROM user_completed_quizzes
+      WHERE id_quiz = ${quizId}`
+    )
+    res.status(200).send(retrieveQuizGlobalRankings);
```
Comment (john-cooke832) [inline server/index.js:198]: If you're sending a 200 status, you can just res.send, which will automatically send a 200.

File: server/index.js [hunk 8]
```diff
@@ -188,6 +188,39 @@ app.get(`/quiz/history/taken/:userId`, async (req, res) => {
...
+      `SELECT *
+      FROM user_completed_quizzes
+      WHERE id_quiz = ${quizId}
+      AND id_users
+      IN
+        (SELECT id_user_friend
+        FROM user_friend_relationships
+        WHERE id_user = ${userId})`
+    )
+    res.status(200).send(retrieveQuizFriendRankings);
```
Comment (john-cooke832) [inline server/index.js:218]: status again

**Review by john-cooke832 — CHANGES_REQUESTED** [review]
A lot of this looks really good. I especially like how explicit your error messages are. I did have a few small notes/requests, though.

File: client/components/takeQuiz/TakeQuiz.jsx (hunk 2 above)
Comment (cindyryoo7) [reply client/components/takeQuiz/TakeQuiz.jsx:192]: This was before Helen's component to pass down the userId was merged to the branch. Fixing it now

**Comment by cindyryoo7** [comment]
@john-cooke832 Made the changes you requested. Please review and approve the PR if everything looks good.

**Review by john-cooke832 — APPROVED** [review]
Looks great, Cindy!!

**Pull request merged**